### PR TITLE
A1

### PR DESCRIPTION
**Pull request by ces-hungtran**
@ces-ThaoPhan review giúp mình, cảm ơn bạn

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
+{
+	"name": "a1-react-training",
+	"version": "1.0.0",
+	"description": "A1 React Training",
+	"devDependencies": {
```
Comment (ces-ThaoPhan) [inline package.json:5]: Please integrate Prettier: https://dev.to/suprabhasupi/learn-to-configure-eslint-and-prettier-in-react-4gp0

File: src/components/BeneficiaryProfile.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import PropTypes from "prop-types";
+import { Formik, Form } from "formik";
+import TextField from "@material-ui/core/TextField";
+import "yup-phone";
+import { getDefaultString, getDefaultDate } from "../utils";
+
+import * as FORM_LABELS from "../constants/formLabels";
+import BENEFICIARY_PROFILE_SCHEMA from "../constants/BeneficiaryProfileSchema";
+
+class BeneficiaryProfile extends React.Component {
```
Comment (ces-ThaoPhan) [inline src/components/BeneficiaryProfile.js:11]: Please convert your components to stateless components. Currently, our React team is following this new style, and React team will support more of this style (performance ...).
We decided on the very 1st meeting that we will only use stateless components, please follow team's decisions, thanks

File: src/components/BeneficiaryProfile.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  render() {
+    return (
+      <Formik
+        initialValues={{
+          hisEmail: getDefaultString(this.props.beneficiaryProfile.hisEmail),
```
Comment (ces-ThaoPhan) [inline src/components/BeneficiaryProfile.js:16]: - just simply `email: this.state.hisEmail || ''`
- I don't think we need to use possessive pronouns when naming variables. It makes the name longer and hard to follow. Please update.

File: src/components/BeneficiaryProfile.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+import { getDefaultString, getDefaultDate } from "../utils";
+
+import * as FORM_LABELS from "../constants/formLabels";
+import BENEFICIARY_PROFILE_SCHEMA from "../constants/BeneficiaryProfileSchema";
+
+class BeneficiaryProfile extends React.Component {
+  render() {
+    return (
+      <Formik
+        initialValues={{
```
Comment (ces-ThaoPhan) [inline src/components/BeneficiaryProfile.js:15]: You could create a init object and use it here

**Comment by ces-ThaoPhan** [comment]
@ces-hungtran Please update, I'll give more comments when you finish those comments above, thanks.

File: package.json (hunk 1 above)
Comment (ces-hungtran) [reply package.json:5]: > Please integrate Prettier: https://dev.to/suprabhasupi/learn-to-configure-eslint-and-prettier-in-react-4gp0

What line do you think is the problem? I've installed eslint + prettier

File: src/components/BeneficiaryProfile.js (hunk 2 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:11]: > Please convert your components to stateless components. Currently, our React team is following this new style, and React team will support more of this style (performance ...). We decided on the very 1st meeting that we will only use stateless components, please follow team's decisions, thanks
So we will use another way to store the state? like react context?

File: src/components/BeneficiaryProfile.js (hunk 3 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:16]: > * just simply `email: this.state.hisEmail || ''`
> * I don't think we need to use possessive pronouns when naming variables. It makes the name longer and hard to follow. Please update.

This is used to distinguish between email in step 1 and step 2, to let the field more specific and meaningful.

File: src/components/BeneficiaryProfile.js (hunk 4 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:15]: > You could create a init object and use it here

and put it into a constants/ file?

**Comment by ces-hungtran** [comment]
@ces-ThaoPhan đã chuyển qua function based component. Review giúp mình với. Cảm ơn bạn

File: src/components/BeneficiaryProfile.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+function BeneficiaryProfile(props) {
+  return (
+    <Formik
+      initialValues={getBeneficiaryProfileInitializeValue(props)}
+      validateOnChange="true"
+      validationSchema={BENEFICIARY_PROFILE_SCHEMA}
+      onSubmit={(value, action) => {
+        const submitValue = Object.assign({}, value);
+        delete submitValue.step;
```
Comment (ces-ThaoPhan) [inline src/components/BeneficiaryProfile.js:18]: Don't need to delete before re-setting for object value

File: src/components/BeneficiaryProfile.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import { Formik, Form } from "formik";
+import TextField from "@material-ui/core/TextField";
+import * as FORM_LABELS from "../constants/formLabels";
+import getBeneficiaryProfileInitializeValue, {
+  BENEFICIARY_PROFILE_SCHEMA,
+} from "../constants/BeneficiaryProfileSchema";
+
+function BeneficiaryProfile(props) {
+  return (
+    <Formik
```
Comment (ces-ThaoPhan) [inline src/components/BeneficiaryProfile.js:12]: Please use `useFormik` hook to have clean code. Using hooks with functional components is best practice. Please update other components.
Read more: https://formik.org/docs/tutorial

File: src/components/BeneficiaryProfile.js (hunk 3 above)
Comment (ces-ThaoPhan) [reply src/components/BeneficiaryProfile.js]: Your data is set up like this 
```
{
 step1: {
  field1: '',
  field2: '',
  ...
 },
 step2: { ... }
}
```
It means fields are specific in their data section, no need to use possessive pronouns. And without possessive pronouns, we could read/find the field names better.

File: src/components/FormInsurance.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
+import React, { useEffect, useState } from "react";
+import BeneficiaryProfile from "./BeneficiaryProfile";
+import FinishCheck from "./FinishCheck";
+import InsuranceProfile from "./InsuranceProfile";
+import FormInsuranceNavigation from "./FormInsuranceNavigation";
+
+import RENDER_STATE from "../constants/renderState";
+
+function FormInsurance(props) {
+  const [state, setState] = useState({
```
Comment (ces-ThaoPhan) [inline src/components/FormInsurance.js:10]: Using Redux or Context API to manage your states globally: (I recommend using Redux)
- You could access states in components directly (no need to pass props)
-  Clean coding

File: src/constants/BeneficiaryProfileSchema.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import * as VALIDATOR_MSG from "./formValidationMessage";
```
Comment (ces-ThaoPhan) [inline src/constants/BeneficiaryProfileSchema.js:1]: This is non-component file, its name has to be camel case `beneficiaryProfileSchemas.js`.
Update others please.

File: src/constants/InsuranceProfileSchema.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import * as VALIDATOR_MSG from "./formValidationMessage";
+import * as Yup from "yup";
+import { getDefaultString, getDefaultDate } from "../utils";
+
+export default function getInsuranceProfileInitializeValue(props) {
+  return {
+    myEmail: getDefaultString(props.insuranceProfile, "myEmail"),
```
Comment (ces-ThaoPhan) [inline src/constants/InsuranceProfileSchema.js:7]: Don't need `getDefaultString`, just simply `email: props.insuranceProfile.email || ''`

**Comment by ces-ThaoPhan** [comment]
I will leave more comments when you have updates, thanks

File: package.json (hunk 1 above)
Comment (ces-hungtran) [reply package.json]: > I couldn't see `prettier` dependency in you `package.json` fiile

I insalled prettier globally (maybe) and add new config file to each project.

File: src/components/BeneficiaryProfile.js (hunk 5 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:18]: > 

If we don't delete a filed, how can we remove it from the state object? set to undefined?

File: src/components/BeneficiaryProfile.js (hunk 6 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:12]: > Please use `useFormik` hook to have clean code. Using hooks with functional components is best practice. Please update other components. Read more: https://formik.org/docs/tutorial

In there getting started, what they use is formik component to demonstration. Is there any benefit here when switch to useFormik?

File: src/components/FormInsurance.js (hunk 7 above)
Comment (ces-hungtran) [reply src/components/FormInsurance.js:10]: > Using Redux or Context API to manage your states globally: (I recommend using Redux)
> 
> * You could access states in components directly (no need to pass props)
> * Clean coding

What I've read from redux docs, they tell me if I can use state, no need to use redux to speed up performace etc...
Do they really make clean code? Is it necessary here? 
Since these states only need to manage in a specific 1, 2 components and we don't share it with the other ones, is there any other benefit than clean code here?

File: src/components/BeneficiaryProfile.js (hunk 3 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js]: > Your data is set up like this
> 
> ```
> {
>  step1: {
>   field1: '',
>   field2: '',
>   ...
>  },
>  step2: { ... }
> }
> ```
> 
> It means fields are specific in their data section, no need to use possessive pronouns. And without possessive pronouns, we could read/find the field names better.

So what I've understood here is "read/find names better" It looks like style-bias here. Anw, let me make the change.

File: src/components/BeneficiaryProfile.js (hunk 6 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js:12]: > 

From what i've read so far, useFormik is good for something simple, for someone want to implement almost everything that Formik conver. And that is what CES are current doing?

File: src/constants/BeneficiaryProfileSchema.js (hunk 8 above)
Comment (ces-hungtran) [reply src/constants/BeneficiaryProfileSchema.js:1]: > This is non-component file, its name has to be camel case `beneficiaryProfileSchemas.js`. Update others please.

Ya, my mistake. Let me change it.

File: src/constants/InsuranceProfileSchema.js (hunk 9 above)
Comment (ces-hungtran) [reply src/constants/InsuranceProfileSchema.js:7]: > Don't need `getDefaultString`, just simply `email: props.insuranceProfile.email || ''`

The props.insuranceProfile also needs checking to avoid any undefined reference here. Like props.undefined.underfined.

File: src/components/BeneficiaryProfile.js (hunk 2 above)
Comment (ces-hungtran) [reply src/components/BeneficiaryProfile.js]: > I will leave more comments when you have updates, thanks

Can you help me by answering my concern above? Thanks a lot.

**Comment by ces-hungtran** [comment]
> I will leave more comments when you have updates, thanks


@ces-ThaoPhan  Can you help me by answering my concern above? Thanks a lot.

File: src/components/BeneficiaryProfile.js (hunk 5 above)
Comment (ces-ThaoPhan) [reply src/components/BeneficiaryProfile.js:18]: Why do you need to remove it? If I got your idea here, you just want to overwrite the value, right?

File: src/components/BeneficiaryProfile.js (hunk 6 above)
Comment (ces-ThaoPhan) [reply src/components/BeneficiaryProfile.js:12]: I recommended before that our React team currently uses functional components and hooks. It's our best practice now. Read React FAQ: https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both.
And please notice that what I comment for you here is what our React team is using now

File: src/components/FormInsurance.js (hunk 7 above)
Comment (ces-ThaoPhan) [reply src/components/FormInsurance.js:10]: I agree with you that your application is not big enough to use Redux. I just want you to learn many ways to manage states in React app and keep your components independent. Imagine when your component is used in many places in your app, you have to pass params every time it is called, instead of using global states.
If you don't really want to learn Redux, you could use Context API.

File: src/constants/InsuranceProfileSchema.js (hunk 9 above)
Comment (ces-ThaoPhan) [reply src/constants/InsuranceProfileSchema.js]: So you need to use optional chaining `props.insuranceProfile?.email || ''`
Tips: https://www.sitepoint.com/shorthand-javascript-techniques/
